### PR TITLE
TRUNK-5903: Switching from hibernate mappings to annotations-Role

### PR DESCRIPTION
**Pull request by mherman22**
## Description of what I changed

- I have changed the role domain from using hibernate xml mappings to annotations.

- Also ensured the role class extends the extends BaseOpenmrsObject but also implements the retireable and Auditable interfaces.
- I made the name attribute transient so that it doesn't get persisted.

## Issue I worked on

see https://issues.openmrs.org/browse/TRUNK-5903

## Checklist: I completed these to help reviewers :)

- [x] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [x] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [x] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command

- [x] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [x] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

### CODE REVIEW

**Comment by tendomart** [comment]
@mherman22  Got some failing tests at

Error:  Tests run: 29, Failures: 0, Errors: 1, Skipped: 2, Time elapsed: 2.353 s <<< FAILURE! - in org.openmrs.hl7.HL7ServiceTest
`Error:  org.openmrs.hl7.HL7ServiceTest.resolveUserId_shouldReturnNullForAmbiguousUsersUsingFirstAndLastNameGivenUserIDIsNull  Time elapsed: 0.06 s  <<< ERROR!`

**Comment by mherman22** [comment]
> @mherman22 Got some failing tests at
> 
> Error: Tests run: 29, Failures: 0, Errors: 1, Skipped: 2, Time elapsed: 2.353 s <<< FAILURE! - in org.openmrs.hl7.HL7ServiceTest `Error: org.openmrs.hl7.HL7ServiceTest.resolveUserId_shouldReturnNullForAmbiguousUsersUsingFirstAndLastNameGivenUserIDIsNull Time elapsed: 0.06 s <<< ERROR!`
 @tendomart i am looking into that. Feel free to read through this talk post though -> https://talk.openmrs.org/t/switching-from-xml-mappings-to-annotations/39277/1

**Comment by mherman22** [comment]
> @mherman22 Got some failing tests at
> 
> Error: Tests run: 29, Failures: 0, Errors: 1, Skipped: 2, Time elapsed: 2.353 s <<< FAILURE! - in org.openmrs.hl7.HL7ServiceTest `Error: org.openmrs.hl7.HL7ServiceTest.resolveUserId_shouldReturnNullForAmbiguousUsersUsingFirstAndLastNameGivenUserIDIsNull Time elapsed: 0.06 s <<< ERROR!`

 @tendomart i am looking into that. Feel free to read through this talk post though -> https://talk.openmrs.org/t/switching-from-xml-mappings-to-annotations/39277/1

File: api/src/main/java/org/openmrs/Role.java [hunk 1]
```diff
@@ -25,20 +37,49 @@
...
 	private String role;
 	
+	@Transient
+	@Field
+	private String name;
+	
+	@Column(name = "description", length = 255)
+	private String description;
+	
+	@ManyToMany(fetch = FetchType.EAGER)
```
Comment (ibacher) [inline api/src/main/java/org/openmrs/Role.java:61]: I think this doesn't correctly port the cascade settings.

File: api/src/main/java/org/openmrs/Role.java [hunk 2]
```diff
@@ -72,11 +127,17 @@ public void setPrivileges(Set<Privilege> privileges) {
...
 	
-	@Override
 	public String getName() {
 		return this.getRole();
 	}
 	
+	/**
+	 * @param name the name to set
+	 */
+	public void setName(String name) {
```
Comment (ibacher) [inline api/src/main/java/org/openmrs/Role.java:137]: What's the point of having this setter and keeping the `name` field around if we neither return the name from `getName()` nor store the field? (I understand keeping `setName()` around, just have it do nothing).

File: api/src/main/java/org/openmrs/Role.java (hunk 1 above)
Comment (ibacher) [reply api/src/main/java/org/openmrs/Role.java:61]: Similarly, there seem to be some cache settings on the M-t-M relations I don't see replicated...

**Review by ibacher — DISMISSED** [review]
Thanks @mherman22! A few small things, but this seems like the right direction...

File: api/src/main/java/org/openmrs/Role.java (hunk 1 above)
Comment (mherman22) [reply api/src/main/java/org/openmrs/Role.java:61]: > I think this doesn't correctly port the cascade settings.

what would be the ideal in this case? i have looked around for a solution that ensures that there's no cascading here

**Comment by mherman22** [comment]
> ```
> Error:    VisitValidatorTest.validate_shouldFailIfTheStartDatetimeIsAfterAnyEncounter » StackOverflow
> Error:    VisitValidatorTest.validate_shouldFailIfTheStopDatetimeIsBeforeAnyEncounter » StackOverflow
> ```

looks to be a failure on the entire core as of this commit --> https://github.com/openmrs/openmrs-core/commit/7e9a3b692d4b913e6c028791bf803b96af59651e

**Comment by ibacher** [comment]
It's a flaky test... Or at least, the commit you point to only updates the README.md file, so I'm not certain how that could change test results.

File: api/src/main/java/org/openmrs/Role.java (hunk 1 above)
Comment (ibacher) [reply api/src/main/java/org/openmrs/Role.java]: I think that's what we actually want (i.e., deleting a role should not delete child roles, inherited roles, or privileges), so maybe just omitting it is fine?

Comment (mherman22) [reply api/src/main/java/org/openmrs/Role.java]: Awesome, lemi omit

**Comment by mherman22** [comment]
> It's a flaky test... Or at least, the commit you point to only updates the README.md file, so I'm not certain how that could change test results.

This test passes when i build locally.

**Comment by mherman22** [comment]
@dkayiwa @ibacher am pinging you on this

File: api/src/main/java/org/openmrs/Role.java [hunk 3]
```diff
@@ -25,20 +38,51 @@
...
 	
-	// Fields
-	
+	@Id
+	@Column(name = "role")
 	private String role;
 	
+	@Transient
+	@Field
+	private String name;
```
Comment (ibacher) [inline api/src/main/java/org/openmrs/Role.java:56]: So I'm still kind of curious why we need this?

Comment (mherman22) [reply api/src/main/java/org/openmrs/Role.java:56]: so this was to make sure that the [name field](https://github.com/openmrs/openmrs-core/blob/df241b5827c76661d1aa769ca39d7c8d8ee06f35/api/src/main/java/org/openmrs/BaseOpenmrsMetadata.java#L35-L37C22) doesnot get persisted in the role table where we only require role,uuid and description.

Comment (ibacher) [reply api/src/main/java/org/openmrs/Role.java:56]: Right, but Role now extends `BaseOpenmrsObject` which doesn't have that field any more right?

Comment (ibacher) [reply api/src/main/java/org/openmrs/Role.java:56]: (I do think this should implement the `OpenmrsMetadata` interface, just with `getName()` and `setName()` modifying the role).

Comment (mherman22) [reply api/src/main/java/org/openmrs/Role.java:56]: > Right, but Role now extends `BaseOpenmrsObject` which doesn't have that field any more right?

oh yeah, my bad




> (I do think this should implement the `OpenmrsMetadata` interface, just with `getName()` and `setName()` modifying the role).

remember this? -> https://talk.openmrs.org/t/switching-from-xml-mappings-to-annotations/39277/4?u=mherman22

Comment (ibacher) [reply api/src/main/java/org/openmrs/Role.java:56]: Well, sure. What I actually care about is ensuring Role implements the `Auditable` and `Retireable` interfaces.

Comment (mherman22) [reply api/src/main/java/org/openmrs/Role.java:56]: awesome, so what i gather here is that role should implement `OpenmrsMetadata` interface and that this ticket -> https://openmrs.atlassian.net/browse/RESTWS-910 is not required

**Comment by dkayiwa** [comment]
@mherman22 did you see the merge conflicts?

**Pull request closed without merging**